netty/src/main/java/io/grpc/netty/NettyClientStream.java

```diff
@@ -105,113 +115,133 @@ public void start(ClientStreamListener listener) {
+                // Remove the bytes from outbound flow control, optionally notifying
+                // the client that they can send more bytes.
+                transportState().onSentBytes(numBytes);
+              }
+            }), flush);
+      } else {
+        // The frame is empty and will not impact outbound flow control. Just send it.
+        writeQueue.enqueue(new SendGrpcFrameCommand(transportState(), bytebuf, endOfStream), flush);
+      }
+    }
 
-  /**
-   * Intended to be overriden by NettyClientTransport, which has more information about failures.
-   * May only be called from event loop.
-   */
-  protected abstract Status statusFromFailedFuture(ChannelFuture f);
+    @Override
+    public void request(int numMessages) {
+      if (channel.eventLoop().inEventLoop()) {
+        // Processing data read in the event loop so can call into the deframer immediately
+        transportState().requestMessagesFromDeframer(numMessages);
+      } else {
+        writeQueue.enqueue(new RequestMessagesCommand(transportState(), numMessages), true);
+      }
+    }
 
-  @Override
-  public void request(int numMessages) {
-    if (channel.eventLoop().inEventLoop()) {
-      // Processing data read in the event loop so can call into the deframer immediately
-      requestMessagesFromDeframer(numMessages);
-    } else {
-      writeQueue.enqueue(new RequestMessagesCommand(this, numMessages), true);
+    @Override
+    public void cancel(Status status) {
+      writeQueue.enqueue(new CancelClientStreamCommand(transportState(), status), true);
     }
   }
 
-  @Override
-  public Integer id() {
-    return id;
-  }
+  /** This should only called from the transport thread. */
+  public abstract static class TransportState extends Http2ClientStreamTransportState
+      implements StreamIdHolder {
+    private final NettyClientHandler handler;
+    private Integer id;
+    private Http2Stream http2Stream;
 
-  public void id(int id) {
-    this.id = id;
-  }
+    public TransportState(NettyClientHandler handler, int maxMessageSize) {
+      super(maxMessageSize);
+      this.handler = checkNotNull(handler, "handler");
+    }
 
-  /**
-   * Sets the underlying Netty {@link Http2Stream} for this stream. This must be called in the
-   * context of the transport thread.
-   */
-  public void setHttp2Stream(Http2Stream http2Stream) {
-    checkNotNull(http2Stream, "http2Stream");
-    checkState(this.http2Stream == null, "Can only set http2Stream once");
-    this.http2Stream = http2Stream;
-
-    // Now that the stream has actually been initialized, call the listener's onReady callback if
-    // appropriate.
-    onStreamAllocated();
-  }
+    @Override
+    public Integer id() {
+      return id;
+    }
 
-  /**
-   * Gets the underlying Netty {@link Http2Stream} for this stream.
-   */
-  @Nullable
-  public Http2Stream http2Stream() {
-    return http2Stream;
-  }
+    public void id(int id) {
```

Comment:
for consistency. We seem to mix `setValue(Object)` and just `value(Object)`. Does it make sense to stick to one style?